test/subdomainregistrar/TestForeverSubdomainRegistrar.js

```diff
@@ -0,0 +1,331 @@
+    FixedPricerERC20 = await deploy('FixedPricer', 1, Erc20.address)
+    FixedPricerFree = await deploy('FixedPricer', 0, EMPTY_ADDRESS)
+  })
+
+  beforeEach(async () => {
+    result = await ethers.provider.send('evm_snapshot')
+  })
+  afterEach(async () => {
+    await ethers.provider.send('evm_revert', [result])
+  })
+
+  describe('setupDomain()', () => {
+    beforeEach(async () => {
+      parentDuration = 86400 * 2
+      await BaseRegistrar.register(labelhash('test'), account, parentDuration)
+      await BaseRegistrar.setApprovalForAll(NameWrapper.address, true)
+      await NameWrapper.wrapETH2LD(
+        'test',
+        account,
+        CANNOT_UNWRAP,
+        EMPTY_ADDRESS,
+      )
+      ;[, , parentExpiry] = await NameWrapper.getData(node)
+      expect(await NameWrapper.ownerOf(node)).to.equal(account)
+    })
+
+    it('should emit an event when a domain is setup', async () => {
+      await expect(
+        SubdomainRegistrar.setupDomain(
+          node,
+          FixedPricer.address,
+          account,
+          true,
+        ),
+      )
+        .to.emit(SubdomainRegistrar, 'NameSetup')
+        .withArgs(node, FixedPricer.address, account, true)
+    })
+  })
+
+  describe('register', () => {
+    let parentExpiry
+
+    beforeEach(async () => {
+      await BaseRegistrar.register(labelhash('test'), account, parentDuration)
+      await BaseRegistrar.setApprovalForAll(NameWrapper.address, true)
+      await NameWrapper.wrapETH2LD(
+        'test',
+        account,
+        CANNOT_UNWRAP,
+        EMPTY_ADDRESS,
+      )
+      ;[, , parentExpiry] = await NameWrapper.getData(node)
+      expect(await NameWrapper.ownerOf(node)).to.equal(account)
+      await SubdomainRegistrar.setupDomain(
+        node,
+        FixedPricerERC20.address,
+        account,
+        true,
+      )
+      await NameWrapper.setApprovalForAll(SubdomainRegistrar.address, true)
+    })
+    it('should allow subdomains to be created', async () => {
+      const balanceBefore = await Erc20WithAccount2.balanceOf(account2)
+
+      const [, fee] = await FixedPricerERC20.price(
+        namehash('test.eth'),
+        'subname',
+        0,
+      )
+
+      await Erc20WithAccount2.approve(
+        SubdomainRegistrar.address,
+        ethers.constants.MaxUint256,
+      )
+
+      await SubdomainRegistrar2.register(
+        node,
+        'subname',
+        account2,
+        EMPTY_ADDRESS,
+        0,
+        [],
+      )
+
+      const balanceAfter = await Erc20WithAccount2.balanceOf(account2)
+      expect(balanceBefore.sub(balanceAfter)).to.equal(fee)
+      const [owner, fuses, expiry] = await NameWrapper.getData(subNode)
+
+      expect(owner).to.equal(account2)
+      expect(expiry).to.equal(parentExpiry - GRACE_PERIOD)
+      expect(fuses).to.equal(CAN_EXTEND_EXPIRY | PARENT_CANNOT_CONTROL)
+    })
+
+    it('should not allow subdomains to be registerd over another domain', async () => {
```

Comment:
Do you mean 'should not allow subdomains to be registered twice`, or `should not allow subdomains to be registered when already registered` ? Wasn't quite sure what `over another domain` mean.